src/index.ts

```diff
@@ -0,0 +1,122 @@
+import {
+	AbstractSqlField,
+	AbstractSqlModel,
+	AbstractSqlTable,
+	InNode,
+} from '@balena/abstract-sql-compiler';
+import { sqlNameToODataName } from '@balena/odata-to-abstract-sql';
+import { replaceResultTransformer, TemplateTag } from 'common-tags';
+
+const trimNL = new TemplateTag(
+	replaceResultTransformer(/^[\r\n]*|[\r\n]*$/g, ''),
+);
+
+const modelNameToCamelCaseName = (s: string) =>
+	s
+		.split(/[ -]/)
+		.map((p) => p[0].toLocaleUpperCase() + p.slice(1))
+		.join('');
+
+const sqlTypeToTypescriptType = (
+	m: AbstractSqlModel,
+	f: AbstractSqlField,
+	opts: RequiredOptions,
+) => {
+	if (!['ForeignKey', 'ConceptType'].includes(f.dataType) && f.checks) {
+		const inChecks = f.checks.find(
+			(checkTuple): checkTuple is InNode => checkTuple[0] === 'In',
+		);
+		if (inChecks) {
+			const [, , ...allowedValues] = inChecks;
+			return allowedValues
+				.map(([type, value]) => (type === 'Text' ? `'${value}'` : value))
+				.join(' | ');
+		}
+	}
+
+	switch (f.dataType) {
+		case 'Boolean':
+			return 'boolean';
+		case 'Short Text':
+		case 'Text':
+		case 'Hashed':
+			return 'string';
+		case 'Date Time':
+			return opts.mode === 'read' ? 'DateString' : 'Date';
+		case 'Serial':
+		case 'Integer':
+		case 'Big Integer':
+		case 'Real':
+			return 'number';
+		// TODO: ForeignKey/ConceptType should really be that of the referenced id
+		case 'ForeignKey':
+		case 'ConceptType':
+			if (opts.mode === 'write') {
+				return 'number';
+			}
```

Comment:
I have a type helper in the SDK which generated the "write" variant given the "read" one.
Imo it's more convenient that having to use two separate types.
See: https://github.com/balena-io/balena-sdk/blob/master/typings/pinejs-client-core.d.ts#L350-L352

Reply:
It's not possible to generate the "write" typings purely from the "read" typings, for instance `Date | string` is valid for write but the read typing is just `string`